### PR TITLE
Extraction files with docs

### PR DESCRIPTION
**Pull request by bharat-suri**
All the scripts used for extracting the text files from the XML dump, finding the surface forms and replacing the anchor text are included in this PR. The scripts are from last year's project and have been cleaned. I have added the documentation to the methods wherever I saw fit.

### CODE REVIEW

**Comment by bharat-suri** [comment]
Hardware config:
Processor 1.6 GHz Intel Core i5
Memory 8 GB

File: gsoc2018-bharat/src/CheckPerson.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+					title = next(inp).rstrip('\n').replace(' ', '_')
+					continue
+				else:
+					wc += len(line.split())
+					for form in mascForms:
+						mc += len(re.findall(r'\b'+ form + r'\b', line, re.IGNORECASE))
+					for form in femiForms:
+						fc += len(re.findall(r'\b'+ form + r'\b', line, re.IGNORECASE))
+				if line.startswith('</doc>'):
+					if (mc/wc > 0.02) or (fc/wc > 0.02):
```
Comment (mommi84) [inline gsoc2018-bharat/src/CheckPerson.py]: Better declare a constant at the beginning of the file and assign it to `0.02`.

File: gsoc2018-bharat/src/CheckPerson.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import os
+import re
+import sys
+
+def countPronouns(filename, f):
```
Comment (tgalery) [inline gsoc2018-bharat/src/CheckPerson.py]: really nice to have docstrings in the code :-) , but there are some other python style issues that would be worth complying with. Filenames are all in lowercase, e.g. `check_person.py` and function names are in snake case `count_pronouns`, snake case is only reserved for class names.

File: gsoc2018-bharat/src/CheckPerson.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	f : Output file with all the titles and
+	their respective form ( Masculine / Feminine ).
+	"""
+	title = ''
+	wc = 0
+	mascForms = ['he', 'him', 'his']
+	mc = 0
+	femiForms = ['she', 'her', 'hers']
+	fc = 0
+	with open(f, 'w+') as output:
```
Comment (tgalery) [inline gsoc2018-bharat/src/CheckPerson.py]: Tabs are strangely formatted, it's usually 4 spaces (PEP-8 assumes whitespaces instead of tabs for indentation) and many modern IDEs (for example intellij) or text editors (check out the anaconda linter if you use sublime text) can help you with that. variable names should also be snake case `masc_forms`. It also might be useful to isolate the operations you do for a single line here as its own function (taking the line string as input). Your code would be more composable and easier to test.

File: gsoc2018-bharat/src/CheckPerson.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+							print(title + ',male')
+						else:
+							output.write(title + ',female')
+							print(title + ',female')
+					wc = mc = fc = nc = pc = 0
+
+if __name__ == "__main__":
+	directory, f = sys.argv[1], sys.argv[2]
+	for root, dirs, files in os.walk(directory):
+		for file in files:
```
Comment (tgalery) [inline gsoc2018-bharat/src/CheckPerson.py]: General note about your code. `file` is a reserved word in many languages, might be easier to select a more appropriate var name, like `wikidump_file` or something.

File: gsoc2018-bharat/src/MakeDictionary.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import sys
+import re
+import os
+import codecs
+
+def upcase_first_letter(s):
+	return s[0].upper() + s[1:]
+
+# Replaces the Gensim class LineSentence
+class MySentences(object):
```
Comment (tgalery) [inline gsoc2018-bharat/src/MakeDictionary.py]: Better name for the corpus iterator ? WikiDescriptions or something like that ?

File: gsoc2018-bharat/src/MakeDictionary.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	the text is treated as the surface form for that very entity.
+	"""
+	counter = 0
+	print("Opened ", directory)
+	sentences = MySentences(directory)
+	dictionary = {}
+	for line in sentences:
+		counter += 1
+
+		# Finding all the anchor text, that is the <a> tags
```
Comment (tgalery) [inline gsoc2018-bharat/src/MakeDictionary.py]: Isolate the operations you do to a line in its own function.

File: gsoc2018-bharat/src/WikiDetector.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
+import os
+import re
+import sys
+import time
+import tempfile
+from joblib import Parallel, delayed
+from urllib.parse import unquote
+from collections import Counter
+
+def upcase_first_letter(s):
```
Comment (tgalery) [inline gsoc2018-bharat/src/WikiDetector.py]: isn't this just `s.capitalize()` ? note that your method would trown an Exception for empty strings.

File: gsoc2018-bharat/src/WikiDetector.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	Arguments
+	---------
+	filename : Input file containing the extracted text.
+	"""
+	print(filename)
+
+	# A temporary file to track the input file document by document.
+	t = tempfile.NamedTemporaryFile(mode = "r+")
+	dictionary = {}
+	with open(filename, 'r') as fil:
```
Comment (tgalery) [inline gsoc2018-bharat/src/WikiDetector.py]: this code block is a bit too long. Might be better to isolate that into separate functions, I'm not entirely sure why you need globals either

File: gsoc2018-bharat/src/combine.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
+import os, sys
```
Comment (tgalery) [inline gsoc2018-bharat/src/combine.py]: It's not great to use commenting / uncommenting as a way to control the logic of your program, might be good to use other resources

File: gsoc2018-bharat/src/lstm.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import argparse
+
+torch.manual_seed(1)
+
+logging.basicConfig(format='%(levelname)s : %(message)s', level=logging.INFO)
+logging.root.level = logging.INFO
+
+wv = None
+
+def load_model(savedModel):
```
Comment (tgalery) [inline gsoc2018-bharat/src/lstm.py]: instead of declaring `wv` as a global, couldn't you just return it ?

File: gsoc2018-bharat/src/lstm.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	model = FastText.load(savedModel)
+	wv = model.wv
+	del model
+
+def load_mappings(descriptions):
+	"""
+	Input to an LSTM must be a 3D tensor. For that, I
+	am reading all the descriptions line by line and
+	storing their embeddings row wise into a numpy
+	array of 3 dimensions and shape (n, 500, 300).
```
Comment (tgalery) [inline gsoc2018-bharat/src/lstm.py]: no need to declare globals if your function just returns them

File: gsoc2018-bharat/src/package/wikifil.pl [hunk 12]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/perl
+
```
Comment (tgalery) [inline gsoc2018-bharat/src/package/wikifil.pl]: Do you need this file ?

File: gsoc2018-bharat/src/package/extract.py [hunk 13]
```diff
@@ -0,0 +1,3238 @@
+#!/usr/bin/env python
```
Comment (tgalery) [inline gsoc2018-bharat/src/package/extract.py]: Do you need this file ? If you are just using this as a preprocessing tool, it might be better to specify this https://github.com/attardi/wikiextractor as a dependency

File: gsoc2018-bharat/src/pre-train.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+from gensim.models.word2vec import LineSentence
+from gensim.models import FastText
+from gensim.models import Word2Vec
+"""
+We are using the wrapper for FastText available in the Gensim package. FastText is an extension to the word2vec
+skipgram model, as it uses character n-grams to generate word embeddings. This allows for the model to create embeddings
+for out of vocabulary words with the help of previously identified n-gram embeddings.
+"""
+logging.basicConfig(format='%(levelname)s : %(message)s', level=logging.INFO)
+logging.root.level = logging.INFO
```
Comment (tgalery) [inline gsoc2018-bharat/src/pre-train.py]: Do you need this, I though you already had pre trained fast text vectors.

**Review by tgalery — CHANGES_REQUESTED** [review]
Lots of comments related to style and code organisation. In general we need a requirements.txt file specifying all the libs you need and a readme.md with detailed instructions on (i) how to do download the data, (ii) where that should be placed, (iii) the order in which we need to run the cli commands to produce the intermediary files that will eventually produce your model

File: gsoc2018-bharat/src/CheckPerson.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
+import os
```
Comment (tgalery) [inline gsoc2018-bharat/src/CheckPerson.py]: If this or other file has been copied and pasted from last years work, there's the question of how the repository should be organised. Should we fix the scripts from the other year , copy and paste, or re organise into a common structure ? @mommi84 might be the person to talk to about this.

Comment (mommi84) [reply gsoc2018-bharat/src/CheckPerson.py]: The best would be to reorganize the code into a common structure. Instead of a folder for each student, they could be named based on the projects, with an additional folder containing common files. Something like:

```
├── evaluation
├── methods
│   ├── baselines (gsoc-nausheen)
│   ├── nlp-based (commons)
│   │   ├── kbed-oov (gsoc-bharat)
│   │   ├── kbed-rva (gsoc-akshay)
```

File: gsoc2018-bharat/src/CheckPerson.py (hunk 2 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/CheckPerson.py]: I will update the scripts accordingly.

File: gsoc2018-bharat/src/CheckPerson.py (hunk 3 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/CheckPerson.py]: Yes, I will see to it that the PEP-8 standards are met with.

File: gsoc2018-bharat/src/CheckPerson.py (hunk 4 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/CheckPerson.py]: Will keep that in mind.

File: gsoc2018-bharat/src/MakeDictionary.py (hunk 5 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/MakeDictionary.py]: Okay

File: gsoc2018-bharat/src/MakeDictionary.py (hunk 6 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/MakeDictionary.py]: Yes that makes sense. I will clean the code a bit more.

File: gsoc2018-bharat/src/WikiDetector.py (hunk 7 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/WikiDetector.py]: The purpose was also to replace spaces with underscore. I will make the changes so that it works inline itself.

File: gsoc2018-bharat/src/combine.py (hunk 9 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/combine.py]: No. That does not alter the flow of control. Nor is it associated with the logic. It was just to count the unique resources in the entire corpus. It also includes the resources that do not have a description but are mentioned nonetheless.

File: gsoc2018-bharat/src/lstm.py (hunk 10 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/lstm.py]: I was doing that originally. I wanted to see the effect it would have on the memory usage. The entire model takes up a big chunk of memory itself. It was better to get rid of it as soon as the vectors were mapped.

File: gsoc2018-bharat/src/lstm.py (hunk 11 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/lstm.py]: Yes, will do that.

File: gsoc2018-bharat/src/pre-train.py (hunk 14 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/pre-train.py]: Pre-trained embeddings for Word2Vec, FastText and GloVe are anyway available. We can use them. This script is to see how the training of these models works in the Gensim library.

File: gsoc2018-bharat/src/package/extract.py (hunk 13 above)
Comment (bharat-suri) [reply gsoc2018-bharat/src/package/extract.py]: Cool. Will add the link to the README and remove the extract.py script from the repo.

**Comment by bharat-suri** [comment]
I added the method for generating the embeddings for entities and their respective descriptions in the train_lstm.py script itself because the use of numpy.stack from the earlier script was causing the process to crash. Now, I am able to work with larger training sets. Latest test was run with 34k descriptions for 10 epochs which took 1 hr of training time.

File: gsoc2018-bharat/README.md [hunk 16]
```diff
@@ -21,13 +32,24 @@ $ wget -c http://mattmahoney.net/dc/enwik9.zip -P data
 $ unzip data/enwik9.zip -d data
 ```
 
-This is a raw Wikipedia dump and needs to be cleaned because it contains a lot of HTML/XML data. There are two ways to pre-process it. Here, I am using the [wikifil.pl](https://github.com/tramplingWillow/ComplexEmbeddings/blob/master/src/package/wikifil.pl) bundled with FastText (the script was originally developed by Matt Mahoney, and can be found on his [website](http://mattmahoney.net/).) to pre-process it.
+This is a raw Wikipedia dump and needs to be cleaned because it contains a lot of HTML/XML data. There are two ways to pre-process it. Here, I am using the Wiki Extractor script available [here](https://github.com/attardi/wikiextractor/blob/master/WikiExtractor.py) bundled with FastText to pre-process it.
 
 ```shell
-$ perl src/package/wikifil.pl data/enwik9 > data/fil9
+$ wget -c https://github.com/attardi/wikiextractor/blob/master/WikiExtractor.py -P src
```
Comment (tgalery) [inline gsoc2018-bharat/README.md:38]: Can't we install this via pip ?

File: gsoc2018-bharat/README.md [hunk 17]
```diff
@@ -41,7 +63,13 @@ The script, [pre-train.py](https://github.com/tramplingWillow/ComplexEmbeddings/
 
 ```shell
 $ mkdir model
-$ python src/pre-train.py -i data/fil9 -o model/pre_wiki -s 300 -sg 1 -hs 1 -e 5
```
Comment (tgalery) [inline gsoc2018-bharat/README.md:44]: might be good to explain what the cli opts are here.

File: gsoc2018-bharat/README.md (hunk 16 above)
Comment (bharat-suri) [reply gsoc2018-bharat/README.md:38]: No, they don't seem to mention any such thing in their README.

File: gsoc2018-bharat/src/Encoder.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
+import torch
```
Comment (tgalery) [inline gsoc2018-bharat/src/Encoder.py:1]: files should never be upercase

File: gsoc2018-bharat/src/Encoder.py [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        seq_len : Length of input sequence
+        num_layers : Number of layers in the LSTM
+        """
+        super(DescriptionEncoder, self).__init__()
+        self.embed_size = embed_size
+        self.hidden_size = hidden_size
+        self.seq_len = seq_len
+        self.num_layers = num_layers
+        self.inp = nn.Linear(embed_size, hidden_size)
+        # self.gru = nn.GRU(hidden_size, hidden_size, 2)
```
Comment (tgalery) [inline gsoc2018-bharat/src/Encoder.py]: maybe remove the commented GRU layer

File: gsoc2018-bharat/src/Encoder.py [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+        self.rnn = nn.RNN(hidden_size // 2, hidden_size,
+                          num_layers, batch_first=True)
+        self.out = nn.Linear(hidden_size, embed_size)
+
+    def forward(self, inp_desc, hidden=None):
+        if hidden is None:
+            hidden = self.init_hidden()
+        inp_desc = self.inp(inp_desc)
+        # output, hidden = self.rnn(self.lstm(inp_desc)[0])
+        # output, hidden = self.rnn(self.lstm(self.gru(inp_desc)[0])[0])
```
Comment (tgalery) [inline gsoc2018-bharat/src/Encoder.py]: same here

File: gsoc2018-bharat/src/create_mappings.py [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+                # Padding the array to a fixed shape.
+                t[:r.shape[0], :r.shape[1]] = r
+                entities.append(v)
+                abstracts.append(t)
+                count += 1
+            except (IndexError, KeyError) as _:
+                continue
+    logging.info('resources read into np stack of length : {0}'.format(count))
+
+    # Currently the process is killed here because of memory pressure.
```
Comment (tgalery) [inline gsoc2018-bharat/src/create_mappings.py]: if this is not a true description of the current state of the code maybe you can remove it

File: gsoc2018-bharat/src/create_mappings.py [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+            # Resource is extracted
+            ent = [_ for _ in json_line.keys()][0]
+
+            # Description is extracted and split into tokens
+            desc = json_line[ent].split()
+            try:
+                # Using a temporary array to pad the original embedding
+                # can be replaced by np.pad()
+                t = np.zeros(shape, dtype=np.float32)
+                v = np.array(wv.get_vector(ent), dtype=np.float32)
```
Comment (tgalery) [inline gsoc2018-bharat/src/create_mappings.py:44]: just wondering if `.get_vector`  could return a `None` and this would cause some of the exceptions you expect below.

File: gsoc2018-bharat/src/create_mappings.py [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+                r = np.array(list(map(lambda x: wv.get_vector(x), desc)),
+                             dtype=np.float32)
+
+                # Padding the array to a fixed shape.
+                t[:r.shape[0], :r.shape[1]] = r
+                entities.append(v)
+                abstracts.append(t)
+                count += 1
+            except (IndexError, KeyError) as _:
+                continue
```
Comment (tgalery) [inline gsoc2018-bharat/src/create_mappings.py]: Maybe do some logging /printing here to show what is going on and add some visibility?

File: gsoc2018-bharat/src/pre-train.py [hunk 24]
```diff
@@ -0,0 +1,99 @@
+import argparse
```
Comment (tgalery) [inline gsoc2018-bharat/src/pre-train.py]: python might have some issues with hyphenated files, for example there might be issues with importing functions or classes from that file and so on. Might be better to rename this `pre_train.py`

File: gsoc2018-bharat/src/train_lstm.py [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+                d = np.array(list(map(lambda x: wv.get_vector(x), desc)),
+                             dtype=np.float32)
+                e = np.array(wv.get_vector(ent.replace('resource/', '')),
+                             dtype=np.float32)
+                entities.append(e)
+                abstracts.append(d)
+            except (IndexError, KeyError) as _:
+                continue
+
+    # logging.info('resources read into np stack : {0}'.format(count))
```
Comment (tgalery) [inline gsoc2018-bharat/src/train_lstm.py]: maybe remove comments if they are not used

**Review by tgalery — APPROVED** [review]
Minor comments about leaving commented code out, file names and some other minor stuff. Afaik you can merge this.
